### PR TITLE
Cleanup code that checks for updates

### PR DESCRIPTION
**Pull request by donatello**
## Description

This patch introduces a new type `updateParams` to parameterize the
code that checks for updates.

* Eliminate passing variables as-is across multiple function (e.g. the
  `mode` string in the earlier code).

* Remove trivial functions (e.g. earlier version of `DownloadReleaseData()` and
  `IsSourceBuild()`).

* Maintain testability of the code.

## Motivation and Context

Cleanup for better readability and maintainability.

## How Has This Been Tested?

Existing test cases pass (`make test`)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by donatello** [comment]
Fixed conflicts. @krishnasrinivas @balamurugana Please review.

File: cmd/update-main.go
Comment (nitisht) [inline cmd/update-main.go]: Probably a better error message here will be `Error in reading cgroup file` ?

Comment (donatello) [reply cmd/update-main.go]: Improved it - we dont have to make it specific to the cgroup file as the check may change in the future.

File: cmd/update-main.go
Comment (krishnasrinivas) [inline cmd/update-main.go]: `updateParams` does not look like the right name for this.
i.e For example why is cgroupFile part of updateParams
why is isDockerInternal a method on updateParams

If the structure name and method names are changed so that they look related then it will look better

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: `version` is better than `versionStr`.  `version` is anyway `string` type.

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: Why pointer is returned than value?

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: `tokens` or `envTokens` is better than `uaPieces`

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: How about `updateChecker` than `updateParams`?

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: `isDockerEnv()` is better than `isDockerInternal()`.

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: IMO this method needs to be exported ie `GetUpdateInfo()`.  This will give better readability at consumer side.  but its your call.

**Review by balamurugana — CHANGES_REQUESTED** [review]
Minor comments

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: Added more comments - this set of parameters enables writing tests for various pieces of logic. It should be more readable now.

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: Changed to the more explicit `updateCheckParams` :)

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: :+1:

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: Refactored a bit more and eliminated this function.

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: :+1:

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: :+1: for envTokens

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: Changed it to not depend on the struct - since the struct can be initialized here. No need to export I think unless we are following a particular convention.

File: cmd/server-main.go
Comment (balamurugana) [inline cmd/server-main.go]: Now I don't see any improvement/fixes the PR does than another way to implement with a struct.

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: cmd/server-main.go
Comment (donatello) [reply cmd/server-main.go]: < moved comment to main-thread so as to not lose it when/if the commit is modified >

**Comment by donatello** [comment]
> Now I don't see any improvement/fixes the PR does than another way to implement with a struct.

@balamurugana This PR is only trying to improve the understandability of the code. It is not trying to fix any bugs. Let me try to list the improvements that are being attempted.

The struct is helping for the following reasons: 

### Avoid threading the `mode` argument through 5 functions

Excerpts from existing code:

```go
func getUpdateInfo(timeout time.Duration, mode string) (older time.Duration, downloadURL string, err error) {...}
func getLatestReleaseTime(timeout time.Duration, mode string) (releaseTime time.Time, err error) {...}
func DownloadReleaseData(timeout time.Duration, mode string) (data string, err error) {...}
func downloadReleaseData(releaseChecksumURL string, timeout time.Duration, mode string) (data string, err error) {...}
func getUserAgent(mode string) string {...}
```

Guess how many of the above functions actually use `mode` than just pass it around?

Just one.


### Avoid trivial functions to just hide and pass-in arguments to functions

1. `IsSourceBuild()`
```go
func isSourceBuild(minioVersion string) bool {
	_, err := time.Parse(time.RFC3339, minioVersion)
	return err != nil
}

func IsSourceBuild() bool {
	return isSourceBuild(Version)
}
```

2. `GetCurrentReleaseTime()`

```go
func GetCurrentReleaseTime() (releaseTime time.Time, err error) {
	return getCurrentReleaseTime(Version, os.Args[0])
}
```

3. `DownloadReleaseData()`

```go
func DownloadReleaseData(timeout time.Duration, mode string) (data string, err error) {
	return downloadReleaseData(minioReleaseURL+"minio.shasum", timeout, mode)
}

```

In this change, I am eliminating these multiple similarly named functions.


### Remove excessive string concatenation

In addition to changes related to introducing the struct, `getUserAgent()` is refactored to be significantly more readable (and less importantly, to have better performance).

```go
// Old
func getUserAgent(mode string) string {
	userAgent := "Minio (" + runtime.GOOS + "; " + runtime.GOARCH
	if mode != "" {
		userAgent += "; " + mode
	}
	if IsDocker() {
		userAgent += "; docker"
	}
	if IsSourceBuild() {
		userAgent += "; source"
	}
	userAgent += ") " + " Minio/" + Version + " Minio/" + ReleaseTag + " Minio/" + CommitID

	return userAgent
}

// New
func (u *updateCheckParams) getUserAgent() string {
	envTokens := []string{runtime.GOOS, runtime.GOARCH}
	if u.serverMode != "" {
		envTokens = append(envTokens, u.serverMode)
	}
	if u.isDocker() {
		envTokens = append(envTokens, "docker")
	}
	if u.isSourceBuild() {
		envTokens = append(envTokens, "source")
	}
	envInfo := strings.Join(envTokens, "; ")
	return fmt.Sprintf("Minio (%s)  Minio/%s Minio/%s Minio/%s",
		envInfo, Version, ReleaseTag, CommitID)
}
```

I hope the points above better illuminate the improvements. They are already mentioned briefly in the commit message. I hope your argument is not that these changes are too trivial to be included.

**Comment by balamurugana** [comment]
> I hope the points above better illuminate the improvements.

:+1: 


> They are already mentioned briefly in the commit message. I hope your argument is not that these changes are too trivial to be included.

The commit message talks about a type `updateParams` which is not exposed/used now other than `getUpdateInfo()`.  You would need to fix that

File: cmd/server-main.go
Comment (balamurugana) [inline cmd/server-main.go]: As `mode` and `timeout` are required to be passed to `getUpdateInfo()`, No point in changing the order.

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: the mid-path name `updateCheckParams` looks odd to me.  As `getUpdateInfo()` is not a method anymore, `updateParams` is enough.

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: Having these methods to pointer is incorrect.

File: cmd/update-main_test.go
Comment (balamurugana) [inline cmd/update-main_test.go:107]: You can add unit tests for `getUserAgent()` and `getDownloadURL()` for codecov more.

**Comment by donatello** [comment]
> The commit message talks about a type `updateParams` which is not exposed/used now other than `getUpdateInfo()`. You would need to fix that

Done.

File: cmd/update-main.go
Comment (balamurugana) [inline cmd/update-main.go]: You could group methods of a type together for better readability.

**Comment by donatello** [comment]
@krishnasrinivas Could you continue with your review?

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: Why is it incorrect?

Comment (balamurugana) [reply cmd/update-main.go]: You can try below
```go
package main

import (
	"fmt"
)

type mytype struct {
	name string
}

func (t mytype) hello() {
	fmt.Println("hello", t.name)
}

func (t *mytype) welcome() {
	fmt.Println("welcome", t.name)
}

func main() {
	t1 := mytype{"world"}
	t1.hello()
	t1.welcome()

	var t2 *mytype
	t2.hello()
	t2.welcome()
}
```

File: cmd/server-main.go
Comment (donatello) [reply cmd/server-main.go]: Done.

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: No, I think making the name explicit is good - it just needs to be read as `update-check params` meaning "parameters used for doing an update-check". The comment also makes this clear.

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: In the above code sample, `t2` is not assigned, so it crashes. Seems like an argument to not have pointer receivers in the Go language! Not sure what your point is, because in my code, the struct is initialized in only place in the application (and few more times in tests), in the same file and not used elsewhere. Obviously it cannot be a nil pointer (by trivial inspection).

Doing some research on this, I found review guidelines in the [Go Wiki](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) - some cherry-picked points that seem relevant (emphasis my own):

> * If the receiver is a large struct or array, a pointer receiver is more efficient. How large is large? Assume it's equivalent to passing all its elements as arguments to the method. If that feels too large, it's also too large for the receiver.
> * If the receiver is a small array or struct that is naturally a value type (for instance, something like the time.Time type), with no mutable fields and no pointers, or is just a simple basic type such as int or string, a value receiver makes sense. A value receiver can reduce the amount of garbage that can be generated; if a value is passed to a value method, an on-stack copy can be used instead of allocating on the heap. (The compiler tries to be smart about avoiding this allocation, but it can't always succeed.) **Don't choose a value receiver type for this reason without profiling first.**
> * Finally, when in doubt, use a pointer receiver.
> 

Though this is not performance sensitive, for curiosity sake, I did some profiling too using this code https://play.golang.org/p/PAuPwIWojW - it does not run in play.golang.org (takes too long), but running on my laptop I get:

```
$ ./testprog 
PointerReceiverResult
2000000000	         0.41 ns/op
ValueReceiverResult
500000000	         3.05 ns/op
```

The program uses a struct that is identical to `updateCheckParams` (has 5 string properties). Clearly the pointer receiver method is faster by like 7X. In summary, it is ok to use pointer receivers here.

File: cmd/update-main.go
Comment (donatello) [reply cmd/update-main.go]: Done.

File: cmd/update-main_test.go
Comment (donatello) [reply cmd/update-main_test.go:107]: Done.

**Comment by donatello** [comment]
@balamurugana @krishnasrinivas Review points are addressed - please take a look again now.

File: cmd/update-main.go
Comment (krishnasrinivas) [inline cmd/update-main.go]: You can notice that all the fields of `updateCheckParams` (except mode) are globally available and hence there is no reason for these fields to be part of this struct.

The intention of code cleanup is correct, the previous code was looking bad i.e mode was being passed all around and in the end to be used by a function somewhere down deep. This can be fixed in a better way by having `globalMinioMode` that gets set during init. Global variables are not a problem as long as no one modifies it, so having mode as globally available is fine. Also Bala is working putting all globals into a single struct so that it looks cleaner.

So you can do away with `updateCheckParams` and have mode globally available.

Comment (donatello) [reply cmd/update-main.go]: Yes, other than `mode` all fields are constants. If you see the previous code, this is clear there too (so why did the earlier code become complex anyway?!) - the reason for putting constant-like values as argument(s) to various functions is to be able to write tests. If these values are constants or are hard-coded then it is not possible to pass different values for the purposes of tests. See for example the test for `isDockerEnv()` - it is quite difficult to write it if the file path string "constant" `/proc/self/cgroup` is hard-coded - but clearly the test is useful and important here and explains why the function is implemented in the way it is (no need to even open `/proc/self/cgroup`!).

The problem with global variables (even those that do not change even during the lifetime of the process) is also related to writing tests. Global values impact multiple tests at the same time and so the test author needs to clearly understand which global variables are used or are affected by a function/code that is to be tested - the test author needs to ensure that before the test they are appropriately set and after the test cleaned up so it doesnt affect subsequent tests. This is generally a pain. This was discussed quite a lot in BLR office some months back with @balamurugana and @krisis - the problem with globals and why they need to be eliminated.

IMO, this is a clean implementation - serving the purpose of maintainability and testability.

Comment (krishnasrinivas) [reply cmd/update-main.go]: There are two options:

1) Passing variables as function parameters so that the functions are unit testable. Downside is longer function signature (or use structs like updateCheckParams and have methods on it)

2) Use global variables, harder to unit test but upside is simpler function signature.

We use 2nd style in minio all over, let's discuss what style we should follow so that we can be consistent.

Comment (donatello) [reply cmd/update-main.go]: @harshavardhana @deekoder Requesting to mark this PR as blocked on @abperiasamy

Comment (krishnasrinivas) [reply cmd/update-main.go]: An issue related to this discussion: https://github.com/minio/minio/issues/3198

File: cmd/update-main.go
Comment (balamurugana) [reply cmd/update-main.go]: * As you told WRT `updateCheckParams` is not performance sensitive. the pointer type is good for methods changing fields of the type.  This case is not applicable to the struct as none of the methods are updating fields.

* As I indicated in the example, pointer type must be initialized to avoid run time panic.  This is avoidable if methods are to value type.

* method to pointer type is even accessible by `nil` pointer.   Having any methods to pointer type will lead to panic.
```go
func (t *mytype) greet() {
	fmt.Println("spock hand")
}

var t2 *mytype
t1.greet()
```

* Choosing pointer or value type for method is by considering performance and/or changing field values.

Comment (donatello) [reply cmd/update-main.go]: > * As you told WRT updateCheckParams is not performance sensitive. the pointer type is good for methods changing fields of the type. This case is not applicable to the struct as none of the methods are updating fields.

Pointer receivers are **required** for methods changing fields of the struct/type. There is no choice in the matter there. The choice is only in the case of methods not updating the struct/type they are attached to.

So what are the considerations in choosing between pointer/value receiver types for such methods? The [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) say that it makes sense for small structs. It also says that you have to think of it as passing all properties of the struct as arguments to each value-receiver method. To me passing 5 variables to so many methods is excessive. So I find a pointer type to be better here. It is not a performance sensitivity argument - but more of an aesthetic to not be wasteful where it is trivial to do so.

You say having pointer receiver has the issue of passing an uninitialized pointer to a method:

> * As I indicated in the example, pointer type must be initialized to avoid run time panic. This is avoidable if methods are to value type.

However, such programmer mistakes are not easily overcome by simply robotically following a convention - see this code:

``` go
package main

import (
	"fmt"
)

type mytype struct{}

func (t mytype) greet() {
	fmt.Println("spock hand")
}

func doSomething(k mytype) {
	k.greet()
}

func main() {
	var t1 *mytype
	doSomething(*t1)
}
```

We have neatly followed the convention and used value-receiver types. However, this also crashes because the caller sent in an unintialized pointer. **The conclusion is that the programmer has to be careful to avoid such pitfalls (especially in Golang). There is no *holy-cow* convention that will help avoid shooting yourself in the foot.**

Again, in this change, the scope of the struct is just one file (ignoring test files for obvious reasons). **It is quite clear that pointers will be always be initialized in this code.** So I still say that it is a fine choice to use pointer-receiver here.


> * method to pointer type is even accessible by nil pointer. Having any methods to pointer type will lead to panic.
> 
> ``` go
> func (t *mytype) greet() {
> 	fmt.Println("spock hand")
> }
> 
> // (fixed a bug here to make it compile)
> var t1 *mytype
> t1.greet()
> ```

Again, I don't understand this point - you say `"Having any methods to pointer type will lead to panic."` but just gave an example yourself of a pointer-receiver method that will [not panic or crash](https://play.golang.org/p/8NKAuImyP1). Here, it appears that it doesn't matter if the method is pointer/value-receiver, because it makes no use of any properties in the type. **Additionally, all the code in the present change are counter-examples to the statement "pointer-receiver methods will lead to panic".**

> * Choosing pointer or value type for method is by considering performance and/or changing field values.

This is not true - as I said earlier, if the method needs to change fields in the struct it has to be a pointer-receiver type. This is compulsory. There is no choice here.

The choice comes only in the case of methods not changing their type's fields. The main consideration in Go coding guidelines *is performance*! Please take some time to go through the [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) - they in fact say to default to pointer types if in doubt.

Comment (krishnasrinivas) [reply cmd/update-main.go]: From [here](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type):

> If the receiver is a large struct or array, a pointer receiver is more efficient. How large is large? Assume it's equivalent to passing all its elements as arguments to the method. If that feels too large, it's also too large for the receiver.

We'll need some sort of convention to decide when a struct is considered "large". We can check with AB on this.

> If the receiver is a small array or struct that is naturally a value type (for instance, something like the time.Time type), with no mutable fields and no pointers, or is just a simple basic type such as int or string, a value receiver makes sense.

This suggests that for anything other than basic types like int, string or a "naturally value type" like time.Time we should use pointer receiver.

> A value receiver can reduce the amount of garbage that can be generated; if a value is passed to a value method, an on-stack copy can be used instead of allocating on the heap. (The compiler tries to be smart about avoiding this allocation, but it can't always succeed.) Don't choose a value receiver type for this reason without profiling first.


_Don't choose a value receiver type for this reason without profiling first._

Very confusing, are they saying use pointers even for basic types like int, string and then perform profiling and then change it to value receivers if there is a performance problem?

Comment (donatello) [reply cmd/update-main.go]: Another official resource on this topic is here - https://golang.org/doc/faq#methods_on_values_or_pointers - appears that value receivers are good (i.e. efficient and clear) for small structs according to the doc.

Also, in some cases, for value receivers, a copy of the struct may end up getting allocated on the heap rather than the stack. This creates more work for the GC as items allocated on the function's call stack are freed at the end of a function call (and don't need the GC). Go determines where to put the allocation based on [Escape Analysis](https://en.wikipedia.org/wiki/Escape_analysis). I couldn't find any official docs about the implementation but there are references to it in golang issue tracker, reddit, mailing list and other places.

**Comment by deekoder** [comment]
@donatello  - Will have a discussion today with @abperiasamy , @krishnasrinivas  and @balamurugana

**Comment by deekoder** [comment]
ping @krishnasrinivas @balamurugana  and @nitisht on this. @donatello can you rebase again.

**Comment by donatello** [comment]
Updated this PR for latest master. This is refactored from scratch, so reviewers may disregard the discussion above up to this point. I have not added a new struct this time, since it was contentious last time.

Some tests are removed - this is following @abperiasamy's advice about not writing code in a weird way simply for the sake of writing unit tests.

**Comment by harshavardhana** [comment]
> Updated this PR for latest master. This is refactored from scratch, so reviewers may disregard the discussion above up to this point. I have not added a new struct this time, since it was contentious last time.

@donatello perhaps you could re-submit the PR to discard the history and freshly review?

**Comment by donatello** [comment]
> @donatello perhaps you could re-submit the PR to discard the history and freshly review?
@harshavardhana ok!

**Pull request closed without merging**
